core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -493,32 +601,76 @@ private Object handleLocalOnlyReadManyCommand(InvocationContext ctx, VisitableCo
+         this.allFuture = allFuture;
+         this.ctx = ctx;
+         this.command = command;
+         this.keys = keys;
+         this.destinationIndex = destinationIndex;
+         this.contactedNodes = contactedNodes;
+         this.helper = helper;
+      }
+
+      @Override
+      public void accept(Map<Address, Response> responseMap, Throwable throwable) {
+         if (throwable != null) {
+            allFuture.completeExceptionally(throwable);
+            return;
+         }
+         Response response = getSingleReadResponseOrFail(responseMap, allFuture);
+         if (response == null) {
+            if (!allFuture.isDone()) {
+               try {
+                  handleMissingResponse();
+               } catch (Throwable t) {
+                  allFuture.completeExceptionally(t);
                }
-            } catch (Throwable t) {
-               allFuture.completeExceptionally(t);
             }
-         });
+            return;
+         }
+         try {
+            Object responseValue = ((SuccessfulResponse) response).getResponseValue();
+            Object[] values = unwrapFunctionalManyResultOnOrigin(ctx, keys, responseValue);
+            if (values != null) {
+               System.arraycopy(values, 0, allFuture.results, destinationIndex, values.length);
+               allFuture.countDown();
+            } else {
+               allFuture.completeExceptionally(new IllegalStateException("Unexpected response value " + responseValue));
+            }
+         } catch (Throwable t) {
+            allFuture.completeExceptionally(t);
+         }
+      }
+
+      private void handleMissingResponse() {
+         Map<Address, List<Object>> requestedKeys = getKeysByOwner(ctx, keys, checkTopologyId(command), null, contactedNodes);
+         int pos = destinationIndex;
+         for (Map.Entry<Address, List<Object>> addressKeys : requestedKeys.entrySet()) {
+            allFuture.increment();
+            List<Object> keysForAddress = addressKeys.getValue();
+            ReplicableCommand remoteCommand = helper.copyForRemote(command, keysForAddress, ctx);
+            Set<Address> target = Collections.singleton(addressKeys.getKey());
+            List<Address> contactedNodes = new ArrayList<>(this.contactedNodes.size() + 1);
+            contactedNodes.addAll(this.contactedNodes);
+            contactedNodes.add(addressKeys.getKey());
+            rpcManager.invokeRemotelyAsync(target, remoteCommand, syncIgnoreLeavers)
+                  .whenComplete(new ReadManyHandler(allFuture, ctx, command, keysForAddress, contactedNodes, pos, helper));
+            pos += keysForAddress.size();
+         }
+         for (int i = destinationIndex + keys.size() - 1; i >= pos; --i) {
+            allFuture.results[i] = LOST_PLACEHOLDER;
```

Comment:
Yes; `null` value is a regular function return value, we can't filter the fixed-size results array for that. And we can't return `null`s for missing entries like nothing happened, since user might expect that the return value from his function is always an instance of `Foo`.